classes/output/login.php

```diff
@@ -49,51 +53,84 @@ public static function output_login_options($wantsurl) {
         }
+        $context->idplist = $idplist;
+
         // If only one IDP is available in authentication plugins then auto-redirect to it.
         $noredirect  = optional_param('noredirect', 0, PARAM_BOOL); // Don't redirect.
         if ($count === 1 && get_config('local_login', 'autoredirect')  && empty($noredirect)) {
             redirect($idploginpath);
         }
 
+        // If data exists in the cache, use it if it is still valid.
+        // We need to check this after calculating the IDP data to ensure its current.
+        $cache = cache::make('local_login', 'renderedlogin');
+        $cachehash = $cache->get('idphash');
+        $idphash = sha1(json_encode($idplist));
+        $cacheval = $cache->get('html');
+        if ($cachehash === $idphash && !empty($cacheval)) {
+            return $cacheval;
+        }
+
         if (!empty($config->showmanual)) {
+            $manual = new \stdClass();
+
             // Now display link to manual login page.
             $urlparams = [
                 'noredirect' => 1,
                 'passive' => 'off' // Prevent Saml2 from triggering passive check on manual login page.
             ];
-            $manualloginurl = new \moodle_url('/login/index.php', $urlparams);
+            $manual->manualurl = new \moodle_url('/login/index.php', $urlparams);
 
-            $output .= \html_writer::start_tag('div', array('class' => 'idp-login container-fluid manual'));
             if (!empty($config->beforemanualtext)) {
-                $output .= \html_writer::span(format_text($config->beforemanualtext), 'beforemanual');
+                $manual->beforemanual = format_text($config->beforemanualtext);
             }
-            $name = !empty($config->custommanualtext) ? format_string($config->custommanualtext)
-            : get_string('manuallogin', 'local_login');
-            $attributes = ['class' => 'btn btn-secondary btn-block', 'title' => $name];
-            $output .= \html_writer::link($manualloginurl, $name, $attributes);
-            $output .= \html_writer::end_tag('div');
+
+            $name = !empty($config->custommanualtext) ? format_string($config->custommanualtext) : get_string('manuallogin', 'local_login');
+            $manual->manualname = $name;
+
+            $context->manual = $manual;
         }
 
-        $output .= \html_writer::end_tag('div');
+        $context->header = format_text($config->headertext);
+        $context->footer = format_text($config->footertext);
 
-        return $output;
+        // Try to load the template from settings.
+        $template = $config->template ?? '';
+        if (empty($template)) {
+            $template = file_get_contents($CFG->dirroot . '/local/login/templates/login-minimal.mustache.default');
+        }
+
+        // Custom mustache engine to load a string into context.
+        $engine = new Mustache_Engine([
+            'escape' => 's',
+            'pragmas' => [\Mustache_Engine::PRAGMA_BLOCKS],
+        ]);
```

Comment:
I assume this is necessary because the templates are saved in config and not to the disk ?
If this is the case, maybe put a comment there explaining that since IMO most devs would scratch their head wondering why you wouldn't just use the core rendering api